pandas/core/indexes/multi.py

```diff
@@ -1412,6 +1415,61 @@ def from_product(cls, iterables, sortorder=None, names=None):
         labels = cartesian_product(labels)
         return MultiIndex(levels, labels, sortorder=sortorder, names=names)
 
+    @classmethod
+    def from_frame(cls, df, squeeze=True):
+        """
+        Make a MultiIndex from a DataFrame.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        squeeze : bool, default True
+            If df is a single column, squeeze multiindex to be a regular index.
+
+        Returns
+        -------
+        MultiIndex or Index
+            The MultiIndex representation of the given DataFrame. Returns an
+            Index if the DataFrame is single column and squeeze is True.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame([[0, 'green'], [0, 'purple'], [1, 'green'],
+        ...                    [1, 'purple'], [2, 'green'], [2, 'purple']],
+        ...                   columns=['number', 'color'])
+        >>> df
+           number   color
+        0       0   green
+        1       0  purple
+        2       1   green
+        3       1  purple
+        4       2   green
+        5       2  purple
+        >>> pd.MultiIndex.from_frame(df)
+        MultiIndex(levels=[[0, 1, 2], ['green', 'purple']],
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 0, 1]],
+                   names=['number', 'color'])
+
+        See Also
+        --------
+        MultiIndex.from_arrays : Convert list of arrays to MultiIndex
+        MultiIndex.from_tuples : Convert list of tuples to MultiIndex
+        MultiIndex.from_product : Make a MultiIndex from cartesian product
+                                  of iterables
+        """
+        from pandas import DataFrame
+        # just let column level names be the tuple of the meta df columns
+        # since they're not required to be strings
+        if not isinstance(df, DataFrame):
+            raise TypeError("Input must be a DataFrame")
+        columns = list(df)
+        mi = cls.from_tuples(list(df.values), names=columns)
+        if squeeze:
+            return mi.squeeze()
+        else:
+            return mi
```

Comment:
Personal nit: change the four lines of `squeeze` logic to one:

~~~
return mi.squeeze() if squeeze else mi
~~~